src/core/i18n/utils.js

```diff
@@ -73,24 +73,93 @@ export function getLanguage(langOrLocale) {
 }
 
 export function isRtlLang(lang) {
-  const language = getLanguage(lang);
+  const language = sanitizeLanguage(lang);
   return rtlLangs.includes(language);
 }
 
 export function getDirection(lang) {
   return isRtlLang(lang) ? 'rtl' : 'ltr';
 }
 
-export function getLangFromRouter(renderProps) {
-  if (renderProps) {
-    // Get the lang from the url param by default
-    // if it exists.
-    if (renderProps.params && renderProps.params.lang) {
-      return getLanguage(renderProps.params.lang);
-    } else if (renderProps.location && renderProps.location.query &&
-               renderProps.location.query.lang) {
-      return getLanguage(renderProps.location.query.lang);
+
+function qualityCmp(a, b) {
+  if (a.quality === b.quality) {
+    return 0;
+  } else if (a.quality < b.quality) {
+    return 1;
+  }
+  return -1;
+}
+
+/*
+ * Parses the HTTP accept-language header and returns a
+ * sorted array of objects. Example object:
+ * {
+ *   lang: 'pl', quality: 0.7
+ * }
+ */
+export function parseAcceptLanguage(header) {
+  // pl,fr-FR;q=0.3,en-US;q=0.1
+  if (!header || !header.split) {
+    return [];
+  }
+  const rawLangs = header.split(',');
+  const langList = rawLangs.map((rawLang) => {
+    const parts = rawLang.split(';');
+    let q = 1;
+    if (parts.length > 1 && parts[1].indexOf('q=') === 0) {
+      const qVal = parseFloat(parts[1].split('=')[1]);
+      if (isNaN(qVal) === false) {
+        q = qVal;
+      }
     }
+    return {
+      lang: parts[0].trim(),
+      quality: q,
+    };
+  });
+  langList.sort(qualityCmp);
+  return langList;
+}
+
+
+/*
+ * Given an accept-language header and a list of currently
+ * supported languages, returns the best match with no normalization.
+ *
+ */
+export function getLangFromHeader(acceptLanguage, { _validLangs } = {}) {
+  let userLang;
+  if (acceptLanguage) {
+    const langList = parseAcceptLanguage(acceptLanguage);
+    for (const langPref of langList) {
+      if (isValidLang(langPref.lang, { _validLangs })) {
+        userLang = langPref.lang;
+        break;
+      // Match locale, even if region isn't supported
+      } else if (isValidLang(langPref.lang.split('-')[0], { _validLangs })) {
+        userLang = langPref.lang.split('-')[0];
+        break;
+      }
+    }
+  }
+  return userLang;
+}
+
+/*
+ * Looks up the language from the router renderProps.
+ * When that fails fall-back to accept-language.
+ *
+ */
+export function getFilteredUserLanguage({ renderProps, acceptLanguage } = {}) {
+  let userLang;
+  // Get the lang from the url param by default if it exists.
+  if (renderProps && renderProps.params && renderProps.params.lang) {
+    userLang = renderProps.params.lang;
+  }
+  // If we don't have a valid userLang set yet try accept-language.
+  if (!isValidLang(userLang) && acceptLanguage) {
```

Comment:
I just mean we should be able to trust that react-router gives us the same `renderProps` structure. So no need to do `renderProps && renderProps.params && renderProps.params.lang`. I think my code is missing a `sanitizeLanguage()` call in the `if`.

It feels like this function is doing a bit too much since it is finding possible langs, validating them and returning a match. I'm not sure if it's worth the refactor but it would be reasonable to me to make this essentially just do `isValidLang(lang) ? sanitizeLang(lang) : defaultLang`.